### PR TITLE
Integration Specs

### PR DESCRIPTION
**Pull request by SaneMyburg**
# Blog app - integration specs for views and fix n+1 problems
### In this PR: 

1. I solved the N+1 problem when fetching all posts and their comments for a user.
     - [Before Screenshot](https://github.com/SaneMyburg/blog-app/assets/109852822/785864af-a75c-4a34-8fba-46ba5b660a41)
     - [After Screenshot](https://github.com/SaneMyburg/blog-app/assets/109852822/11bd9acb-b1d9-4e95-8572-b8af1f39edc5)

    

3. I used  Capybara to write Integration specs as follows:
- For user index page:
  - I can see the username of all other users.
  - I can see the profile picture for each user.
  - I can see the number of posts each user has written.
  - When I click on a user, I am redirected to that user's show page.

- For user show page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see the user's bio.
  - I can see the user's first 3 posts.
  - I can see a button that lets me view all of a user's posts.
  - When I click a user's post, it redirects me to that post's show page.
  - When I click to see all posts, it redirects me to the user's post's index page.

- For user post index page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see a post's title.
  - I can see some of the post's body.
  - I can see the first comments on a post.
  - I can see how many comments a post has.
  - I can see how many likes a post has.
  - I can see a section for pagination if there are more posts than fit on the view.
  - When I click on a post, it redirects me to that post's show page.

- For post show page:
  - I can see the post's title.
  - I can see who wrote the post.
  - I can see how many comments it has.
  - I can see how many likes it has.
  - I can see the post body.
  - I can see the username of each comment.
  - I can see the comment each commentor left.

### CODE REVIEW

File: spec/integration/user_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  before do
+    user
+    visit users_path
+  end
+
+  scenario 'I can see the username, profile picture, and number of posts for each user' do
+    expect(page).to have_content(user.name)
+    expect(page.html).to include(user.photo)
+    expect(page).to have_content(user.posts_counter)
+  end
```
Comment (TedLivist) [inline spec/integration/user_index_spec.rb:18]: - [ ] **[OPTIONAL]** It is often best to separate these tests into separate blocks to provide a more organized test and explanation. You merged three scenarios into one, but I would advise you split them into three.

File: spec/integration/user_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    end
+
+    it 'should have a link to see all posts' do
+      expect(page.html).to include('See All Posts')
+    end
+    it 'should show next_path when clicked on see all posts' do
+      click_on 'See All Posts'
+      expect(current_path).to eq(current_path)
+    end
+  end
```
Comment (TedLivist) [inline spec/integration/user_show_spec.rb]: - [ ] You are missing the test for redirecting to the post's show page when a user clicks on a post. Kindly implement that.

File: spec/integration/post_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    end
+    it 'should show user photo' do
+      expect(page.html).to include(@user1.photo)
+    end
+    it 'should show posts_counter of clicked user' do
+      expect(page).to have_content(@user1.posts_counter)
+    end
+    it 'should show all posts' do
+      expect(page).to have_content(@first_post.title)
+    end
```
Comment (TedLivist) [inline spec/integration/post_index_spec.rb]: - [ ] I don't think the description for this test fits the actual test. It should check for post title, not all posts.

File: spec/integration/post_index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      expect(page).to have_content(@first_post.comments_counter)
+    end
+    it 'should show likes counter' do
+      expect(page).to have_content(@first_post.likes_counter)
+    end
+    it 'should show a button for pagination if there are more posts than fit on the view.' do
+      visit user_posts_path(@user1.id)
+      expect(page).to have_button('Pagination')
+    end
+  end
```
Comment (TedLivist) [inline spec/integration/post_index_spec.rb]: - [ ] You are missing the test for redirecting to post show page when a user clicks on a post. Kindly add that.

File: spec/integration/post_show_spec.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+      expect(page).to have_content(@first_post.comments_counter)
+    end
+
+    it 'should show likes counter' do
+      expect(page).to have_content(@first_post.likes_counter)
+    end
+
+    it 'should show name of commenter' do
+      expect(page).to have_content(@comment1.author.name)
+    end
```
Comment (TedLivist) [inline spec/integration/post_show_spec.rb:41]: - [ ] You are missing tests for the following scenarios:

  - Tests that asserts you can see who wrote the post.

**Review by TedLivist — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ ♻️

Hi @SaneMyburg, @preciousbetine and @Tracy-miranja,

Good job so far!
✔️ Well-written and passing tests
However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the **[OPTIONAL]** prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/integration/post_show_spec.rb (hunk 5 above)
Comment (Davidosky007) [reply spec/integration/post_show_spec.rb:41]: * This test still missing , kindly add it `Tests that asserts you can see who wrote the post.`

**Review by Davidosky007 — CHANGES_REQUESTED** [review]
Hi @SaneMyburg 👋, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
* N+1 issue solved 👍 
* Integration test written with capybara gem 👍 

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @davidosky007 in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Reem-lab — APPROVED** [review]
# Hi Team,  👋🏻👋🏻

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


## Highlights
- [ ] Great job on the funtionality for the tests💯
- [ ] Followed Gitflow 🔥
- [ ] No Linters error ✅
- [ ] All tests are passed 💯💯
- [ ] Kudos wokring on solving the `n+1` problems⭐


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**